### ISSUE TITLE
Would it be possible to resize arrays without reallocating?

### ISSUE THREAD

**Issue by svick** [dotnet/coreclr#11686]
Currently, when a type like `List<T>` wants to resize an array, it has to allocate a new array and copy all the elements from the old array to the new one (possibly using `Array.Resize()`). I think this copying can take a fairly long time, especially for large arrays.

But considering that on 64-bit platforms, there should be enough free virtual address space, would it be possible to "reserve" more address space for such growing arrays, so that they can be grown in place?

The way I imagine it could work is something like this:

* There would be a new type `ResizableArray<T>`*, which would mostly behave just like an array, except that its `Length` can change and that its `Resize` method (`public static void Resize(ref ResizableArray<T> array, int minNewSize)`) either allocates a new `ResizableArray<T>` or changes its `Length` (see below). It has to be a new type, because nobody expects that the `Length` of `T[]` can change.
* If `ResizableArray<T>.Resize` was called with a small `minNewSize`, it would behave similarly to `Array.Resize` and reallocate.
* If called with a sufficiently large `minNewSize`  on a small `ResizableArray<T>` (the limit could be the same as for LOH, or larger, depending on how fast reallocating an array is and how much wasting of virtual address space is acceptable), the array would be reallocated into a special area of memory, reserving its maximum size for it (AFAIK 2<sup>32</sup> · element size) in the virtual address space, but actually allocating only as much physical memory as requested. This would make it a "large" `ResizableArray<T>`.
* If called on a large `ResizableArray<T>`, it would just allocate the new memory, without copying anything, since the virtual address space is available.
* If virtual address space ran out because of too many resizable arrays (e.g. more than ~16k resizable `int` arrays on architectures with 48-bit address space like x64), part of the reserved address space for some resizable array could be reused. Calling `Resize` on that array could then potentially reallocate it.
* Since `ResizableArray<T>.Resize` would actually allocate physical memory from the OS by pages, it could resize the array to more than what was asked for, which would mean no physical memory would be wasted (that's why the parameter is called `minNewSize`).
* Just like normal array, `ResizableArray<T>` could be converted to `Span<T>`, which means there would be a common abstraction.

My questions:

* Would this be possible?
* Could this actually sufficiently increase performance?
* Would the cost of making this work and adding the new type be warranted?
* Are there better ways to solve this issue?

I think creating this type is primarily a matter of implementation, not API, which is why I'm opening this issue here, and not on corefx.

---

\* This name is very similar to [`ResizeArray<T>`](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/collections.resizearray['t]-type-abbreviation-[fsharp]), which is an F# alias for `System.Collections.Generic.List<T>`. So maybe a different name would be better?

**Comment by mikedn** [comment]
> Would this be possible?

Likely. But there's at least one ugly issue: thread safety. The GC needs to know the length of the array to be able to scan it. The JIT needs to know the length the array to be able to eliminate range checks. If the array length can change then things may get complicated.

> Could this actually sufficiently increase performance?

Yes, provided that an application actually makes significant use of this feature (that includes allocating large enough arrays since you say that this feature kicks in after the array length exceeds a certain threshold).

> Would the cost of making this work and adding the new type be warranted?

Do you know of any other language/framework/runtime which has such a feature (excluding the fact that in C/C++ you can implement it yourself)? If the answer is "no" then the answer to your own question is likely "no" as well :)

> Are there better ways to solve this issue?

Yep. Avoid resizing large arrays.

**Comment by svick** [comment]
> But there's at least one ugly issue: thread safety.

Good point, but I think those problems might be solvable.

> The GC needs to know the length of the array to be able to scan it.

Concurrent GC already has to handle object mutations (like assigning references). Maybe a similar kind of write barrier could be used here?

> The JIT needs to know the length the array to be able to eliminate range checks. If the array length can change then things may get complicated.

If the array size could only grow, then eliminated range checks based on old size would still be valid. Considering that types like `List<T>` generally don't shrink the underlying array on their own, I think that would be acceptable.

> Do you know of any other language/framework/runtime which has such a feature?

I don't. That could mean what you imply, but there are also other explanations. For example, it could be that this only makes sense in .Net, because it's fairly unique in caring a lot about safety, convenience *and* performance (see `unsafe`). Or it could be that this is a truly innovative idea (see `await`). Yeah, that last part is not very likely. 🙂 

> Avoid resizing large arrays.

Except that types that do resize potentially large arrays (like `List<T>` and `Dictionary<K, V>`) are very convenient and fairly efficient. Making them even more efficient without any loss of convenience would be great, IMO.

**Comment by ygc369** [comment]
@svick 
Two problems:
1. On 32-bit platforms, there may not be enough free virtual address space, how would your `ResizableArray<T>` behave?
2. If the size reduces, would it free some memory?

**Comment by svick** [comment]
> On 32-bit platforms, there may not be enough free virtual address space, how would your `ResizableArray<T>` behave?

I'm not sure. The type would either be much less useful on 32-bit, or it wouldn't give any advantage over array at all. For now, let's just assume it would behave like array and reallocate for every `Resize`.

> If the size reduces, would it free some memory?

If reducing the size of a `ResizableArray<T>` was actually allowed (see my answers to @mikedn for why it might not), then yes, I think it should deallocate memory. Why would that be a problem?

**Comment by mikedn** [comment]
> Concurrent GC already has to handle object mutations (like assigning references). Maybe a similar kind of write barrier could be used here?

Changing references is a rather different situation. But yes, there are probably solutions to thread safety issues like this. For example when extending the array Resize should:
* commit the memory
* zero it out
* update length

This prevents the GC from scanning into memory that contains garbage. Of course, zeroing memory has a cost but that's unavoidable.

> Considering that types like List<T> generally don't shrink the underlying array on their own, I think that would be acceptable.

Actually being able to decrease the array length can be useful. For example `List<T>.Clear` has to clear the array if it contains references. If we could change the array length to 0 then clearing would not be needed. This allows reusing `List<T>` objects at no cost.

> That could mean what you imply, but there are also other explanations.

Perhaps. My main concern with this approach is that it depends on OS services that maybe are not designed for this. Virtual address space tricks like this are commonly used (the GC itself has the habit of reserving a huge chunk of address space upfront for its heap) but I don't know if anyone attempted to use such tricks on a large scale. We're talking about the possibility of creating thousands of VA allocations at a high rate.

> Except that types that do resize potentially large arrays (like List<T> and Dictionary<K, V>) are very convenient and fairly efficient. Making them even more efficient without any loss of convenience would be great, IMO.

Sure but at least in some cases the reallocation problem can be avoided
* specify a reasonably capacity when creating the collection
* create/use collections that do not store all the data in a single large array
* don't load more data than you need

Sure, avoidance is not always possible but the above choices work in many cases. Possibly many enough that the remaining cases are rare enough that the cost of this feature is just too high.

**Comment by svick** [comment]
> My main concern with this approach is that it depends on OS services that maybe are not designed for this.

What specifically are you worried about? That doing this would degrade performance of those OS services?

But yes, it's possible this would not actually work well. I guess the first step in prototyping this would be to write a native application that simulates how this would behave under load as closely as possible and run it on the supported OSes.

> Possibly many enough that the remaining cases are rare enough that the cost of this feature is just too high.

That's certainly possible.

**Comment by mikedn** [comment]
> What specifically are you worried about? That doing this would degrade performance of those OS services?

Yes. For example the following piece of code takes a rather long time on Windows:
```C++
for (int i = 0; i < 10000; i++)
    VirtualAlloc(nullptr, 2147483648, MEM_RESERVE, PAGE_READWRITE);
```
If the allocation size is 8 times smaller then the same code runs instantly. That's kind of odd and I've seen other such issues.

**Comment by ygc369** [comment]
A good idea, but seems only useful for arrays larger than a page.

**Comment by svick** [comment]
@ygc369 Yeah. In my proposal, `ResizableArray<T>.Resize` would reallocate small arrays and it would actually reserve the address space only for arrays larger than some threshold, likely one that's even more than 4 kB, which is the common size of pages.

**Comment by jnm2** [comment]
@svick The alternative to this is for people to allocate a normal array of the larger size and expose that along with a length and manage everything manually. Leaving it this way seems fine because you'd only need this for targeted performance optimization, right?

**Comment by svick** [comment]
@jnm2 

> The alternative to this is for people to allocate a normal array of the larger size up front and expose that along with a mutable length and manage everything manually.

That's not quite the same:

1. With my proposal, if you create, say, 1 MB `ResiableArray<byte>`, it will reserve 4 GB of virtual memory, but only allocate 1 MB of physical memory. That's not something you can do with a normal array. With that, you have to decide whether you want to waste physical memory or waste time copying. With `ResizableArray<T>`, you don't have to do either.
2. What you're saying requires that you know what is the maximum size of the array. What if you don't know? Do you allocate 4 billion element array? `ResizableArray<T>` does not have this problem.

> Leaving it this way seems fine because you'd only need this for targeted performance optimization, right?

Types like `List<T>` and `Dictionary<K, V>` could use `ResizableArray<T>` instead of normal array. That way, performance of increasing the size of every sufficiently large `List<T>` would improve.
